cdap-tms/src/main/java/io/cdap/cdap/messaging/service/CoreMessagingService.java

```diff
@@ -438,22 +438,27 @@ private Map<String, String> createDefaultProperties() {
     return properties;
   }
 
+  private TopicMetadata getTopic(TopicId topicId) throws TopicNotFoundException, IOException {
+    try {
+      return topicCache.get(topicId);
+    } catch (ExecutionException e) {
+      Throwable cause = Objects.firstNonNull(e.getCause(), e);
+      Throwables.propagateIfPossible(cause, TopicNotFoundException.class, IOException.class);
+      throw Throwables.propagate(e.getCause());
+    }
+  }
+
   @Override
   public CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException {
-    final TopicMetadata metadata =
-        new DefaultTopicMetadata(
-            messageFetchRequest.getTopicId(),
-            getTopicMetadataProperties(messageFetchRequest.getTopicId()));
-    MessageTable messageTable = createMessageTable(metadata);
+    TopicMetadata metadata = getTopic(messageFetchRequest.getTopicId());
+
+    TableProvider<MessageTable> messageTableProvider = () -> createMessageTable(metadata);
```

Comment:
All this provider thing is just complicates it. The provider is called immediately in the `MessageCloseableIterator` constructor, is not it? If so, can we get rid of it as it's just unnecessary complexity. It had sense before with builder when "prepareFetch" and "fetch" were separate